lib/spack/spack/solver/asp.py

```diff
@@ -1889,7 +1920,11 @@ def define_version_constraints(self):
 
             # This is needed to account for a variable number of
             # numbers e.g. if both 1.0 and 1.0.2 are possible versions
-            exact_match = [v for v in allowed_versions if v == versions]
+            exact_match = [
+                v
+                for v in allowed_versions
+                if v == versions and not isinstance(v, spack.version.GitVersion)
```

Comment:
I wouldn't use `==` here to start with, it's like expecting `1 == [1]` (in my versions PR I'm dropping "promotion" of 1 to [1] so it's more clear)

Reply:
`versions` can be a list, range, or a single version. Using e.g. `satisfies` would probably be sufficient for my needs in this PR, but I don't think it would be sufficient for the scenarios it was initially created for (avoid the concretizer picking `1.x.y` instead of `1` if there is an explicit `1` version in the package).

Admittedly nothing fails in `spec_semantics.py`, `spec_dag.py`, `concretize.py`, or `concretize_requirements.py` if I trade `==` for `.satisfies`, which I'd expect to be the modules to cover this change, but again I don't trust that is in fact sensible given what I interpret its intent to be. The commit which adds this at https://github.com/spack/spack/commit/ada2fa36a9aae3b2f328ac5c553c30cf25bd79df doesn't add the test and is part of the much larger https://github.com/spack/spack/pull/19501 (I tried all of the test modules which looked relevant from that PR though).

Reply:
Okay, let's leave it in the questionable state it's in and I'll fix it in my pr.